Spigot-API-Patches/0317-Adding-enhanced-Material-Item-Color-Type-API.patch

```diff
@@ -0,0 +1,594 @@
++
++import java.util.Arrays;
++import java.util.Collections;
++import java.util.Objects;
++import java.util.Set;
++
++/**
++ * Represents different kind of possible Colors, which colorable {@link Material}s also/and {@link DyeableMaterial}s can have
++ */
++public enum ColorType {
++
++    NONE,
++    WHITE,
++    ORANGE,
++    MAGENTA,
++    LIGHT_BLUE,
++    YELLOW,
++    LIME,
++    PINK,
++    GRAY,
++    LIGHT_GRAY,
++    CYAN,
++    PURPLE,
++    BLUE,
++    BROWN,
++    GREEN,
++    RED,
++    BLACK;
++
++    /**
++     * Gets all supported colored {@link Material}s, which can be used
++     * @return all possible colored {@link Material}s
++     */
++    @NotNull
++    public static Set<Material> getAllVariants() {
++        Set<Material> result = new java.util.HashSet<>(Collections.emptySet());
++
++        for (ColorType colorType : values()) {
++            result.addAll(
++                Arrays.asList(
++                    colorType.getWool(),
++                    colorType.getTerracotta(),
++                    colorType.getConcrete(),
++                    colorType.getConcretePowder(),
++                    colorType.getCarpet(),
++                    colorType.getStainedGlass(),
++                    colorType.getStainedGlassPane(),
++                    colorType.getShulkerBox(),
++                    colorType.getGlazedTerracotta(),
++                    colorType.getBed(),
++                    colorType.getBanner(),
++                    colorType.getBannerWall(),
++                    colorType.getDye()
++                )
++            );
++        }
++
++        return result;
++    }
++
++    /**
++     * Gets the {@link ColorType} of an {@link Material}
++     * @param material The {@link Material}, where the {@link ColorType} will come from
++     * @return the {@link ColorType} of the {@link Material} <i>material</i>, or null if it has no supported color
++     */
++    @Nullable
++    public static ColorType getColorType(@NotNull Material material) {
++        Set<Material> allVariants = getAllVariants();
++        if (allVariants.stream().anyMatch(it -> it == material)) {
++            for (Material loopMaterial : allVariants) {
++                if (loopMaterial == material) {
++                    return valueOf(loopMaterial.name().split("_")[0]);
++                }
++            }
++        }
++        return null;
++    }
++
++    /**
++     * Gets this {@link ColorType} color as an {@link DyeColor}
++     * @return the {@link DyeColor} of this {@link ColorType}
++     */
++    @NotNull
++    public DyeColor getDyeColor() {
++        if (this == NONE) {
++            return DyeColor.WHITE;
++        } else
++            return DyeColor.valueOf(name());
++    }
++
++    /**
++     * Gets this {@link ColorType} color as an {@link Color}, maybe for particles
++     * @return the {@link Color} of this {@link ColorType}
++     */
++    @NotNull
++    public Color getParticleColor() {
++        Color result;
```

Comment:
with java 16 we can use a switch expression here, right?, but even with java 8, the cases should be `return Color.whatever;`, not assigning a var and breaking.

Reply:
Yes, it's true, I still had an incorrect JDK loaded, which meant I didn't have the option to do this. I thought I was wrong when I wanted to do it that way, however it was the JDK.